### DISCUSSION TITLE
void::Void?

### DISCUSSION THREAD

**Discussion by jkvargas** [libp2p/rust-libp2p#2812] [Q&A]
So, I got the type for event error which seems to be "
libp2p_swarm::SwarmEvent<blink_impl::behavior::BehaviourEvent, libp2p_core::either::EitherError<libp2p_core::either::EitherError<libp2p_core::either::EitherError<libp2p_core::either::EitherError<libp2p_core::either::EitherError<libp2p_gossipsub::error::GossipsubHandlerError, std::io::error::Error>, std::io::error::Error>, either::Either<libp2p_swarm::handler::ConnectionHandlerUpgrErr<libp2p_core::either::EitherError<libp2p_relay::v2::protocol::inbound_hop::FatalUpgradeError, libp2p_relay::v2::protocol::outbound_stop::FatalUpgradeError>>, void::Void>>, void::Void>, libp2p_ping::handler::Failure>>"

I cannot find out what should I use in place of void::Void, or what is either::Either
Since I want to specify the type in a function signature.
Can you guys help me to figure this out?
Thanks.

**Comment by thomaseizinger** [comment]
Both of them are crates that you can depend on if you really need to name the type!

The crates are:

- `either`
- `void`

But it is usually not necessary to name these types. Can you give more context on what you are actually trying to do?

**Comment by jkvargas** [comment]
basically I want to create a trait that defines a function that accepts an event as a parameter
SwarmEvent<MyBehaviour, ErrorType>

**Reply by elenaf9** [reply]
Does the trait implementation care about what the error type is, i.e. does it handle it? Otherwise you could just make it generic like this:
```rust
trait MyTrait {
    fn handle_swarm_event<E>(&mut self, event: SwarmEvent<MyBehaviour, E>);
}
```
In practice `E` might be a fixed type, but I find using generics easier than complex type aliases.
